rdar://89218473 ([ER] WebKitTestRunner and DumpRenderTree should let the user know if loading a file or a url failed)

mscottapple opened the pull request:
#### 1f07bde4413aea4d24324609b7ef02c5316eb988
<pre>
rdar://89218473 ([ER] WebKitTestRunner and DumpRenderTree should let the user know if loading a file or a url failed)

Reviewed by David Kilzer (ddkilzer).

This adds preflight checking for the existence of input files specified on the command line, as well as reporting of any URL loading errors.

* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(initializeGlobalsFromCommandLineOptions):
(runTest):
* Tools/DumpRenderTree/mac/ResourceLoadDelegate.h:
* Tools/DumpRenderTree/mac/ResourceLoadDelegate.mm:
(-[ResourceLoadDelegate dealloc]):
(-[ResourceLoadDelegate webView:resource:didFailLoadingWithError:fromDataSource:]):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::createOtherPlatformWebView):
(WTR::TestController::createWebViewWithOptions):
(WTR::TestController::createTestURL):
(WTR::TestController::runTest):
(WTR::TestController::didFailProvisionalNavigation):
(WTR::pathSuitableForTestResult):
* Tools/WebKitTestRunner/TestController.h:

Canonical link: <a href="https://commits.webkit.org/268472@main">https://commits.webkit.org/268472@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34cbc2fa91357cb8e26ee1ca3b7797cc569d4cae

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19823 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20247 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20865 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/21715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/18517 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20058 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23505 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/20393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/21715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20043 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/20029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/17237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/22569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/17201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/18032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/22569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/18277 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/18206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/22569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/18799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/20393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/17968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/22316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2419 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/18625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Tools/DumpRenderTree/mac/DumpRenderTree.mm:
```diff
@@ -1915,6 +1919,20 @@ static void runTest(const std::string& inputLine)
+    }
+
+    ResourceLoadDelegate *delegate = (__bridge ResourceLoadDelegate *)resourceLoadDelegate().get();
+    delegate.mainResourceURLString = url.absoluteString;
```
You should not have to explicitly unwrap the `ResourceLoadDelegate` here, or use a bridging cast.

Something like this should work (untested):

```suggestion
    [resourceLoadDelegate() setMainResourceURLString:url.absoluteString];
```

Otherwise if you don't like old-style setters, something like this should work (untested):

```suggestion
    resourceLoadDelegate().get().mainResourceURLString = url.absoluteString;
```

ddkilzer commented on Tools/DumpRenderTree/mac/ResourceLoadDelegate.mm:
```diff
@@ -263,7 +271,16 @@ -(void)webView: (WebView *)wv resource:identifier didFinishLoadingFromDataSource
+        NSString* failingURLString = failingURL.absoluteString;
+        // Allow for an exact match, or a match with the trailing '/' missing from the argument URL
+        if ([mainResourceURLString isEqualToString:failingURLString]
+            || ([failingURLString hasPrefix:mainResourceURLString] && failingURLString.length - mainResourceURLString.length == 1)) {
```
The second test here could result in a false-positive match if you were to compare these two paths:

```
    mainResourceURLString = @"http://localhost/my/path"
    failingURLString  = @"http://localhost/my/pathX"
```

We probably need to check that the leftover character is literally a `/` character.

I'd be really surprised if we didn't have code that did this already.

In fact, you might be able to use `-[NSURL(WebNSURLExtras) _webkit_canonicalize]` or `-_webkit_canonicalize_with_wtf` from `Source/WebKitLegacy/mac/Misc/WebNSURLExtras.h` (since this is in DumpRenderTree which tests WK1, and WebKitLegacy implements WK1).

mscottapple replied:
I was hoping there might be some kind of canonicalization methods—I didn't much care for this check otherwise. (And I noticed that WKURL didn't have this same problem with comparison.)

ddkilzer commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -1498,7 +1507,12 @@ WKURLRef TestController::createTestURL(const char* pathOrURL)
     }
 
-    return WKURLCreateWithUTF8CString(buffer.get());
+    auto cPath = buffer.get();
+    if (!WTF::FileSystemImpl::fileExists(String(cPath + 7, strlen(cPath) - 7))) {
```
Are we guaranteed that `cPath` always starts with `file://` here?  Or do we need a length check before indexing into `cPath`?

mscottapple replied:
Both paths through the prior `if` statement copy "file://" into the beginning of the buffer, so yes, `cPath` will always start with "file://".

ddkilzer commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -1667,7 +1681,12 @@ bool TestController::runTest(const char* inputLine)
+    if (!testURLRef)
+        return false;
+
+    m_mainResourceURL = adoptWK(testURLRef);
```
Usually it's preferable to adopt the object when it's created.  If `createTestURL` returns `nullptr`, the `WKURLRef` will just be empty, but the check on the next line still works:

```suggestion
    auto testURLRef = adoptWK(createTestURL(command.pathOrURL.c_str()));
    if (!testURLRef)
        return false;

    m_mainResourceURL = testURLRef;
```

Also, if there's no reason not to always set `m_mainResourceURL` (even if the value is `nullptr`), then this becomes:

```suggestion
    m_mainResourceURL = adoptWK(createTestURL(command.pathOrURL.c_str()));
    if (!m_mainResourceURL)
        return false;
```

ddkilzer commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -1684,6 +1703,7 @@ bool TestController::runTest(const char* inputLine)
 
     m_currentInvocation->invoke();
     m_currentInvocation = nullptr;
+    m_mainResourceURL = 0;
```
We prefer `nullptr` in these cases where you're clearing out a smart pointer.

```suggestion
    m_mainResourceURL = nullptr;
```

mscottapple replied:
Using `nullptr` was my first choice, but for some reason at some point in development that was generating an error while `0` wasn't. (The ivar was a different type, not `WKRetainPtr<WKURLRef>`, as it is now.)

ddkilzer reviewed (approved):
r=me, but please consider fixing the changes I note below.

ddkilzer commented on Tools/DumpRenderTree/mac/ResourceLoadDelegate.mm:
```diff
@@ -118,6 +119,14 @@ - (NSString *)_drt_descriptionSuitableForTestResult
+
+- (void)dealloc
+{
+    mainResourceURL = nil;
```
Just realized that this will leak because DumpRenderTree is not compiled with ARC.  (Assigning to `nil` is optional without ARC, but makes the switch to ARC easier later as you just delete the `-release` call.)

```suggestion
    [mainResourceURL release];
    mainResourceURL = nil;
```

ddkilzer reviewed (approved):
r=me, we just need to fix the leak.

mscottapple committed:
rdar://89218473 ([ER] WebKitTestRunner and DumpRenderTree should let the user know if loading a file or a url failed)

Reviewed by David Kilzer (ddkilzer).

This adds preflight checking for the existence of input files specified on the command line, as well as reporting of any URL loading errors.

* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(initializeGlobalsFromCommandLineOptions):
(runTest):
* Tools/DumpRenderTree/mac/ResourceLoadDelegate.h:
* Tools/DumpRenderTree/mac/ResourceLoadDelegate.mm:
(-[ResourceLoadDelegate dealloc]):
(-[ResourceLoadDelegate webView:resource:didFailLoadingWithError:fromDataSource:]):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::createOtherPlatformWebView):
(WTR::TestController::createWebViewWithOptions):
(WTR::TestController::createTestURL):
(WTR::TestController::runTest):
(WTR::TestController::didFailProvisionalNavigation):
(WTR::pathSuitableForTestResult):
* Tools/WebKitTestRunner/TestController.h:

Canonical link: https://commits.webkit.org/268472@main
